[Merged by Bors] - Add demo kafka-druid-earthquake-data (#72)
stackabletech/stackablectl: docs/modules/ROOT/pages/demos/kafka-druid-earthquake-data.adoc

The Druid deep storage is based on the S3 store provided by MinIO.

=== View ingestion job
You can have a look at the ingestion job running in Druid by opening the given `druid` endpoint `router-http` from your `stackablectl services list` command output. You have to use the endpoint from your command output, in this case it is http://172.18.0.4:30109. Open it with your favorite browser.

image::demo-kafka-druid-earthquake-data/druid_1.png[]

By clicking on `Ingestion` at the top you can see the running ingestion jobs.

image::demo-kafka-druid-earthquake-data/druid_2.png[]

After clicking on the magnification glass to the right side of the `RUNNING` supervisor you can see additional information.
On the tab `Statistics` on the left you can see the number of processed records as well as the number of errors.
The statistics show that Druid is currently ingesting 1251 records/s and has ingested 2,1 million records so far.
All records have been ingested successfully, which is indicated by having no `processWithError`, `thrownAway` or `unparseable`.

image::demo-kafka-druid-earthquake-data/druid_3.png[]

> === Query datasource
> The started ingestion job has automatically created the Druid datasource `earthquakes`.
> You can see the available datasources by clicking on `Datasources` at the top.

Reviewer: Im pretty sure its data source (when mentioning the druid tab `Datasources` its fine). This is used alot, i would go for a refactor.
```suggestion
=== Query the data source
The started ingestion job has automatically created the Druid data source `earthquakes`.
You can see the available data sources by clicking on `Datasources` at the top.
```

== Summary
The demo streamed 1000 earthquake records/s for a total of ~3 million earthquakes into a Kafka steaming pipeline.
Druid ingested the data near real-time into its data source and enabled SQL access to it.
Superset was used as a web-based frontend to execute SQL statements and build dashboards.

== Where to go from here
There are multiple paths to go from here.
The following sections can give you some ideas on what to explore next.
You can find the description of the earthquake data https://earthquake.usgs.gov/earthquakes/feed/v1.0/csv.php[on the United States Geological Survey website].

=== Execute arbitrary SQL statements
Within Superset (or the Druid webinterface) you can execute arbitrary SQL statements to explore the earthquake data.

=== Create additional dashboards
You also have the possibility to create additional charts and bundle them together in a Dashboard.
Have a look at https://superset.apache.org/docs/creating-charts-dashboards/creating-your-first-dashboard#creating-charts-in-explore-view[the Superset documentation] on how to do that.

=== Load additional data

> You can use a kafka client like https://docs.confluent.io/platform/current/app-development/kafkacat-usage.html[kafkacat] to create new topics and ingest data.

Reviewer: ```suggestion
You can use a Kafka client like https://docs.confluent.io/platform/current/app-development/kafkacat-usage.html[kafkacat] to create new topics and ingest data.
```

= kafka-druid-earthquake-data

[NOTE]
====
This guide assumes you already have the demo `kafka-druid-earthquake-data` installed.
If you don't have it installed please follow the xref:commands/demo.adoc#_install_demo[documentation on how to install a demo].
To put it simply you have to run `stackablectl demo install kafka-druid-earthquake-data`.
====

This demo will

* Install the required Stackable operators
* Spin up the follow data products

> ** *Superset*: A modern data exploration and visualization platform. This demo uses it to execute SQL queries on Druid and build dashboards

Reviewer: ```suggestion
** *Superset*: A modern data exploration and visualization platform. This demo utilizes Superset to retrieve data from Druid via SQL queries and build dashboards on top of that data
```

> ** *Kafka*:  A distributed event streaming platform for high-performance data pipelines, streaming analytics and data integration. This demos uses it as a event streaming platform to stream the data in near-realtime

Reviewer: ```suggestion
** *Kafka*:  A distributed event streaming platform for high-performance data pipelines, streaming analytics and data integration. This demos uses it as a event streaming platform to stream the data in near real-time
```

> ** *Druid*: A real-time database to power modern analytics applications. This demo uses it to ingest the near-realtime data from Kafka, store it and enable SQL access to it

Reviewer: ```suggestion
** *Druid*: A real-time database to power modern analytics applications. This demo uses it to ingest the near real-time data from Kafka, store it and enable to access the data via SQL
```

                                                                                                                     
 kafka      kafka        default    kafka              172.18.0.4:30968                                              
                                                                                                                     
 superset   superset     default    external-superset  http://172.18.0.3:32108  Admin user: admin, password: admin   
                                                                                                                     
 zookeeper  zookeeper    default    zk                 172.18.0.2:32309                                              
                                                                                                                     
 minio      minio-druid  default    http               http://172.18.0.4:32306  Third party service                  
                                    console-http       http://172.18.0.4:31664  Admin user: root, password: rootroot
----

[NOTE]
====
When a product instance has not finished starting yet, the service will have no endpoint.

> Starting all of the product instances might take an considerable amount of time depending on your internet connectivity.

Reviewer: ```suggestion
Starting all of the product instances might take a considerable amount of time depending on your internet connectivity.
```

In case the product is not ready yet a warning might be shown.
====

== Inspect data in Kafka

> Kafka is used as an event streaming platform to stream the data in near-realtime.

Reviewer: ```suggestion
Kafka is used as an event streaming platform to stream the data in near real-time.
```

All the messages put in and read from Kafka are structured in dedicated queues called topics.

> The testdata will be put in a topic called `earthquakes`.

Reviewer: ```suggestion
The test data will be put in a topic called `earthquakes`.
```

> The records get produced (put in) by the testdata-generator and are consumed (read) by Druid afterwards in the same order they where produced.

Reviewer: ```suggestion
The records get produced (put in) by the test data generator and are consumed (read) by Druid afterwards in the same order they where produced.
```

As Kafka itself has no webinterface you have to use a Kafka client like https://docs.confluent.io/platform/current/app-development/kafkacat-usage.html[kafkacat].

> To endpoint where Kafka is reachable is given in the `kafka` endpoint from the `kafka` service in your `stackablectl services list` command output (`172.18.0.4:30968` in this case).

Reviewer: ```suggestion
In order to connect to Kafka, use the `kafka` endpoint from the `kafka` service in your `stackablectl services list` command output (`172.18.0.4:30968` in this case).
```

If you have `kafkacat` installed, you can see the available brokers and topics with the following command.
If not, that's no big deal as you would see nearly the exact same output.
You need to replace the `-b` parameter to match your Kafka endpoint.

[source,console]
----
$ kafkacat -b 172.18.0.4:30968 -L
Metadata for all topics (from broker -1: 172.18.0.4:30968/bootstrap):
 1 brokers:
  broker 1001 at 172.18.0.4:30755 (controller)
 1 topics:
  topic "earthquakes" with 1 partitions:
    partition 0, leader 1001, replicas: 1001, isrs: 1001
----

> You can see the Kafka consists of one broker and the topic `earthquakes` has been created.

Reviewer: ```suggestion
You can see that Kafka consists of one broker and the topic `earthquakes` has been created.
```

----

If you are interested on how many records have been produced to the Kafka topic so far, use the following command.
It will print the last record produced to the topic, which will be formatted with the pattern specified in the `-f` parameter.
The given pattern will print some metadata of the record.

[source,console]
----
$ kafkacat -b 172.18.0.4:30968 -C -t earthquakes -o -1 -c 1 \
    -f 'Topic %t / Partition %p / Offset: %o / Timestamp: %T\n'
Topic earthquakes / Partition 0 / Offset: 2670281 / Timestamp: 1660571644060
----

From the output you can see, that the last record was produced at the timestamp `1660571644060` which translates to `Mo 15. Aug 15:54:04 CEST 2022` (using the command `date -d @1660571644`).
You can also see that it was the record number `2670281` send to this topic, so ~2.6 million million records have been produced so far.

== Druid

> Druid is used to ingest the near-realtime data from Kafka, store it and enable SQL access to it.

Reviewer: ```suggestion
Druid is used to ingest the near real-time data from Kafka, store it and enable SQL access to it.
```

> The demo has started a ingestion job reading earthquake records from the Kafka topic `earthquakes` and saving it into Druids storage.

Reviewer: ```suggestion
The demo has started an ingestion job reading earthquake records from the Kafka topic `earthquakes` and saving it into Druids deep storage.
```

image::demo-kafka-druid-earthquake-data/druid_2.png[]

After clicking on the magnification glass to the right side of the `RUNNING` supervisor you can see additional information.
On the tab `Statistics` on the left you can see the number of processed records as well as the number of errors.
The statistics show that Druid is currently ingesting 1251 records/s and has ingested 2,1 million records so far.
All records have been ingested successfully, which is indicated by having no `processWithError`, `thrownAway` or `unparseable`.

image::demo-kafka-druid-earthquake-data/druid_3.png[]

=== Query datasource
The started ingestion job has automatically created the Druid datasource `earthquakes`.
You can see the available datasources by clicking on `Datasources` at the top.

image::demo-kafka-druid-earthquake-data/druid_4.png[]

By clicking on the `earthquakes` datasource you can see the segments the datasource consists of.
In this case the `earthquakes` datasource is partitioned by the year of the earthquake, resulting in 73 segments.

image::demo-kafka-druid-earthquake-data/druid_5.png[]

Druid offers a web-based way of querying the datasources via SQL.

> To achieve this you first have to on `Query` at the top.

Reviewer: ```suggestion
To achieve this you first have to click on `Query` at the top.
```

image::demo-kafka-druid-earthquake-data/druid_6.png[]

You can now enter any arbitrary SQL statement, to e.g. list 10 earthquakes run

[source,sql]
----
select * from earthquakes limit 10
----

image::demo-kafka-druid-earthquake-data/druid_7.png[]

To count the number of earthquakes per year run

[source,sql]
----
select
  time_format(__time, 'YYYY') as "year",
  count(*) as earthquakes
from earthquakes
group by 1
order by 1 desc
----

image::demo-kafka-druid-earthquake-data/druid_8.png[]

== Superset

> Superset gives the ability to execute SQL queries and build dashboards.

Reviewer: ```suggestion
Superset provides the ability to execute SQL queries and build dashboards.
```

Open the `superset` endpoint `external-superset` in your browser (http://172.18.0.3:32108 in this case).

image::demo-kafka-druid-earthquake-data/superset_1.png[]

Log in with the credentials username `admin`, password `admin`.

image::demo-kafka-druid-earthquake-data/superset_2.png[]

=== View dashboard
The demo has created a Dashboard to visualize the earthquake data.
To open it click on the tab `Dashboards` at the top.

image::demo-kafka-druid-earthquake-data/superset_3.png[]

Click on the dashboard called `Earthquakes`.
It might take some time until the dashboards renders all the included charts.

image::demo-kafka-druid-earthquake-data/superset_4.png[]

=== View charts

The dashboard `Earthquakes` consists of multiple charts.
To list the charts click on the tab `Charts` at the top.

image::demo-kafka-druid-earthquake-data/superset_5.png[]

// TODO: Fix typo in Chart name
Click on the Chart `Number of earthquakes my magnitude`.

> On the left side you can modify the chart and click on `Run` to see the effect.

Reviewer: The typo being capitalization or what?

Author: typo in chart name has been fixed

image::demo-kafka-druid-earthquake-data/superset_6.png[]

=== View distribution on world map

> To look at geographical distribution of the earthquakes you have to click on the tab `Charts` at the top again.

Reviewer: ```suggestion
To look at the geographical distribution of the earthquakes you have to click on the tab `Charts` at the top again.
```

The distribution of the earthquakes matches the continental plate margins.
This is the expected distribution from the https://en.wikipedia.org/wiki/Earthquake[Wikipedia article on Earthquakes].

You can move and zoom the map with your mouse to interactively explore the map.
You can e.g. have a detailed look at the detected earthquakes in Germany.

image::demo-kafka-druid-earthquake-data/superset_8.png[]

You can also click on the magnitudes in the legend on the top right side to enable/disable printing the earthquakes of that magnitude.
By only enabling magnitudes greater or equal to 8 you can only plot the most severe earthquakes.

image::demo-kafka-druid-earthquake-data/superset_9.png[]

=== Execute arbitrary SQL statements
Within Superset you can not only create dashboards but also run arbitrary SQL statements.
On the top click on the tab `SQL Lab` -> `SQL Editor`.

image::demo-kafka-druid-earthquake-data/superset_10.png[]

On the left select the database `druid`, the schema `druid` and set `See table schema` to `earthquakes`.

image::demo-kafka-druid-earthquake-data/superset_11.png[]

On the right textbox enter the desired SQL statement.

> If you do not want to make on up you can use the following:

Reviewer: ```suggestion
If you do not want to make one up, you can use the following:
```

[source,sql]
----
select
  time_format(__time, 'YYYY') as "year",
  count(*) as earthquakes
from earthquakes
group by 1
order by 1 desc
----

image::demo-kafka-druid-earthquake-data/superset_12.png[]

== MinIO
The S3 provided by MinIO is used as a persistent storage for Druid to store all the data used.
Open the `minio` endpoint `console-http` in your browser (http://172.18.0.4:31664 in this case).

image::demo-kafka-druid-earthquake-data/minio_1.png[]

Log in with the credentials username `root`, password `rootroot`.

image::demo-kafka-druid-earthquake-data/minio_2.png[]

Click on the blue button `Browse` on the bucket `druid` and open the folders `data` -> `earthquakes`.

image::demo-kafka-druid-earthquake-data/minio_3.png[]

As you can see druid saved 199MB of data within 73 prefixes (folders).

> On prefix corresponds to on segment which in turn contains all the data of a year.

Reviewer: ```suggestion
One prefix corresponds to on segment which in turn contains all the data of a year.
```

> If you don't see any years or for files the reason is, that Druid has not saved it's data from memory to the deep storage (in this case S3).

Reviewer: ```suggestion
If you don't see any years or files, the reason is that Druid has not saved it's data from memory to the deep storage (in this case S3).
```

After waiting a bit the data should have been flushed to S3 and should show up.

image::demo-kafka-druid-earthquake-data/minio_3.png[]

If you open up a prefix for a specific year you can see that Druid has placed a file containing the data of this year there.

== Summary
The demo streamed 1000 earthquake records/s for a total of ~3 million earthquakes into a Kafka steaming pipeline.

> Druid ingested the data near-realtime into its datasource and enabled SQL access to it.

Reviewer: ```suggestion
Druid ingested the data near real-time into its data source and enabled SQL access to it.
```

Superset was used as a web-based frontend to execute SQL statements and build dashboards.

== Where to go from here
There are multiple paths to go from here.
The following sections can give you some ideas on what to explore next.
You can find the description of the earthquake data https://earthquake.usgs.gov/earthquakes/feed/v1.0/csv.php[on the United States Geological Survey website].

=== Execute arbitrary SQL statements

> Within Superset (or the Druid webinterface) you can execute arbitrary SQL statements to explore the taxi data.

Reviewer: ```suggestion
Within Superset (or the Druid webinterface) you can execute arbitrary SQL statements to explore the earthquake data.
```

=== Create additional dashboards
You also have the possibility to create additional charts and bundle them together in a Dashboard.
Have a look at https://superset.apache.org/docs/creating-charts-dashboards/creating-your-first-dashboard#creating-charts-in-explore-view[the Superset documentation] on how to do that.

=== Load additional data

> You can use a kafka client like https://docs.confluent.io/platform/current/app-development/kafkacat-usage.html[kafkacat] to create new topics and  ingest data.

Reviewer: ```suggestion
You can use a kafka client like https://docs.confluent.io/platform/current/app-development/kafkacat-usage.html[kafkacat] to create new topics and ingest data.
```

> Using the Druid webinterface you can start a ingestion job that consumes the data and stores it in an internal datasource.

Reviewer: ```suggestion
Using the Druid web interface, you can start an ingestion job that consumes the data and stores it in an internal data source.
```

There is a great https://druid.apache.org/docs/latest/tutorials/tutorial-kafka.html#loading-data-with-the-data-loader[tutorial] from Druid on how to do this.

> Afterwards the datasource is available for analysis within Druid and Superset the same way the earthquake data is.

Reviewer: ```suggestion
Afterwards the data source is available to be analyzed within Druid and Superset the same way the earthquake data is.
```
